### PR TITLE
Fix #17017 (math edge cases)

### LINKED ISSUE #17017: Windows: math functions buggy edge cases
`math.sqrt` does not return `NaN` on Windows for negative arguments and instead returns the argument.

### Possible Solution
This seems to be a [MinGW-w64 bug](https://sourceforge.net/p/mingw-w64/bugs/567/) that is fixed by now, so updating the MinGW-w64 version should fix this.

### Additional Information
I actually only tested this via the CI, see https://github.com/konsumlamm/Nim/runs/1883950662.

---

EDIT: Other `math` functions also have edge cases.
* [x] sqrt
* [x] gamma
* [x] c_frexp2

### PR DESCRIPTION
**Pull request by konsumlamm**
fixes #17017
refs https://github.com/nim-lang/Nim/pull/16725#pullrequestreview-590855306

## future work
- [ ] add frexp vmops (timotheecour), cf pr_vmops_frexp

### CODE REVIEW

File: lib/pure/math.nim [hunk 1]
```diff
@@ -949,18 +949,25 @@ func frexp*[T: float32|float64](x: T): tuple[frac: T, exp: int] {.inline.} =
     doAssert frexp(8.0) == (0.5, 4)
     doAssert frexp(-8.0) == (-0.5, 4)
     doAssert frexp(0.0) == (0.0, 0)
+
     # special cases:
-    when not defined(windows):
-      doAssert frexp(-0.0) == (-0.0, 0) # signbit preserved for +-0
-      doAssert frexp(Inf).frac == Inf # +- Inf preserved
-      doAssert frexp(NaN).frac.isNaN
+    doAssert frexp(-0.0) == (-0.0, 0) # signbit preserved for +-0
```
Comment (konsumlamm) [inline lib/pure/math.nim]: This does not actually check for exactly -0.0 (since `0.0 == -0.0`), however, it is good for illustration (and it's not wrong), so I kept it.

Comment (timotheecour) [reply lib/pure/math.nim]: since then we've added signbit; use it here, maybe:
```
doAssert frexp(-0.0)[0].signbit
```
(only for that case; we have tests anyways for other cases in tests/)
(and in future PR, turning doAssert=>assert for all runnableExamples in std/math)

File: tests/stdlib/tmath.nim [hunk 2]
```diff
@@ -370,8 +367,42 @@ template main() =
     doAssert a1.clamp(a2..a4) == a2
     doAssert clamp((3, 0), (1, 0) .. (2, 9)) == (2, 9)
 
-  when not defined(windows): # xxx pending bug #17017
-    doAssert sqrt(-1.0).isNaN
+  block: # edge cases
+    doAssert sqrt(-4.0).isNaN
+
+    doAssert ln(0.0) == -Inf
+    doAssert ln(-0.0) == -Inf
```
Comment (timotheecour) [inline tests/stdlib/tmath.nim:374]: is that what c math does? (instead of NaN)

Comment (timotheecour) [reply tests/stdlib/tmath.nim:374]: EDIT: yes

```
#include <stdio.h>
#include <math.h>
int main (int argc, char *argv[]) {
  double a = -0.0;
  double b = log(a);
  printf("%g\n", b);
  return 0;
}
```
-inf

Comment (konsumlamm) [reply tests/stdlib/tmath.nim:374]: Yes, see also https://en.cppreference.com/w/c/numeric/math/log.

File: lib/pure/math.nim [hunk 3]
```diff
@@ -949,18 +949,25 @@ func frexp*[T: float32|float64](x: T): tuple[frac: T, exp: int] {.inline.} =
...
     var exp: cint
     result.frac = c_frexp2(x, exp)
     result.exp = exp
   else:
     if x == 0.0:
-      result = (0.0, 0)
+      # reuse signbit implementation
+      let uintBuffer = toBitsImpl(x)
+      if (uintBuffer[1] shr 31) != 0:
+        # x is -0.0
```
Comment (timotheecour) [inline lib/pure/math.nim:967]:
```suggestion
      if x.signbit:
```

File: tests/stdlib/tmath.nim [hunk 4]
```diff
@@ -370,8 +367,42 @@ template main() =
...
+
+    doAssert log10(0.0) == -Inf
+    doAssert log10(-0.0) == -Inf
+    doAssert log10(-12.0).isNaN
+
+    doAssert log2(0.0) == -Inf
+    doAssert log2(-0.0) == -Inf
+    doAssert log2(-12.0).isNaN
+
+    when nimvm: discard
```
Comment (timotheecour) [inline tests/stdlib/tmath.nim:385]: i have a followup PR to support this

**Review by timotheecour — APPROVED** [review]
LGTM i can address remaining comments + do the future work i mentioned in a followup PR

File: lib/pure/math.nim (hunk 3 above)
Comment (konsumlamm) [reply lib/pure/math.nim:967]: The problem is that `signbit` is `since (1, 5, 1)`, see also https://github.com/nim-lang/Nim/pull/16725#discussion_r568301079.
